Ad9545 add ref priorities

tachicialex opened the pull request:
Add support for TDC sources priorities.

TDC sources can be RefA, RefB and auxiliary NCOs. AD9545 will use as input for the DPLLs a valid TDC source with the highest priority specified in the DT. If an error occurs on a used TDC source AD9545 will automatically revert to a valid one with a lower priority.

Even if only one source can be used at a time by a DPLL, when setting the rate of the DPLL, the driver will
compute N div, FRAC, MOD for all possible TDC sources. Those parameters are stored on chip in Translation Profiles. See:
ad9545_pll_set_rate

This is needed in order to maintain the same PLL output frequency, even after reference switching.

tachicialex commented:
Changelog v1 -> v2:
- updated yaml example

nunojsa commented on Documentation/devicetree/bindings/clock/clk-ad9545.yaml:
```diff
@@ -167,27 +167,47 @@ patternProperties:
+
+        adi,pll-loop-bandwidth-uhz:
+          description: |
+            PLL loop bandwidth in microherz.
```
typo in hertz...

nunojsa commented on Documentation/devicetree/bindings/clock/clk-ad9545.yaml:
```diff
@@ -167,27 +167,47 @@ patternProperties:
+          allOf:
+            - $ref: /schemas/types.yaml#/definitions/uint32
+            - minimum: 1
+            - maximum: 1850
```
I don't really remember why but AFAICT, but things like this can be defined as (no even sure if the allOf is really ok here):
```
  description: |
￼      PLL loop bandwidth in microherz.
  $ref: /schemas/types.yaml#/definitions/uint32
  minimum: 1
  maximum: 1850
```

Look [here](https://elixir.bootlin.com/linux/latest/source/Documentation/devicetree/bindings/example-schema.yaml#L145) for example (ditto for all other places where this applies)

nunojsa commented on Documentation/devicetree/bindings/clock/clk-ad9545.yaml:
```diff
@@ -167,27 +167,47 @@ patternProperties:
-          - maximum: 1850
-        maxItems: 1
+    "^profile@[0-5]$":
+      description: |
```
nit: Don't think we really need formatted text here...

nunojsa commented on Documentation/devicetree/bindings/clock/clk-ad9545.yaml:
```diff
@@ -303,8 +323,12 @@ examples:
+                               adi,pll-source = <4>;
+                               adi,profile-priority = <0>;
+                               adi,pll-loop-bandwidth-uhz = <200000000>;
+                        };
```
As we don't have any CI for schemas, please make sure that `make dt_binding_check  DT_SCHEMA_FILES=$this_schema` works. You can further  play with this and change the example values to incorrect values to see if your schema is working as expected (eg: adi,pll-source = <10>; should give us an error)...

nunojsa commented on drivers/clk/adi/clk-ad9545.c:
```diff
@@ -314,13 +331,24 @@ struct ad9545_out_clk {
+	unsigned int			address;
+	unsigned int			parent_index;
+	unsigned int			priority;
+	u8				tdc_source;
```
nit: I would place the `u8` and probably the `bool` at the end of the struct (being the u8 the last)

nunojsa commented on drivers/clk/adi/clk-ad9545.c:
```diff
@@ -500,20 +531,41 @@ static int ad9545_parse_dt_plls(struct ad9545_state *st)
+		/* parse DPLL profiles */
+		fwnode_for_each_available_child_node(child, profile_node) {
+			ret = fwnode_property_read_u32(profile_node, "reg", &profile_addr);
+			if (ret < 0)
```
`dev_err()?`

nunojsa commented on drivers/clk/adi/clk-ad9545.c:
```diff
@@ -500,20 +531,41 @@ static int ad9545_parse_dt_plls(struct ad9545_state *st)
 
-		st->pll_clks[addr].loop_bw_uhz = val;
+			ret = fwnode_property_read_u32(profile_node, "adi,pll-loop-bandwidth-uhz", &val);
+			if (ret < 0)
```
ditto...

nunojsa commented on drivers/clk/adi/clk-ad9545.c:
```diff
@@ -1378,8 +1492,15 @@ static unsigned long ad95452_pll_clk_recalc_rate(struct clk_hw *hw, unsigned lon
+	/* if no ref is valid, pll will run in free run mode */
+	i = ad9545_pll_get_parent(hw);
+	if (i == clk->num_parents)
+		return clk->free_run_freq;
```
nit: new line

nunojsa commented on drivers/clk/adi/clk-ad9545.c:
```diff
@@ -1485,33 +1621,54 @@ static int ad9545_plls_setup(struct ad9545_state *st)
+			if (ret < 0)
+				return ret;
+
+			if (pll->profiles[j].tdc_source > 3)
```
What is the 3 and 4 meaning? I would prefere to have some meaningful thing here (as macro or sizeof() - if applicable) as it might make things a bit neater to read

nunojsa reviewed (commented):
Looks good. Mostly things related with bindings

tachicialex commented:
Changelog V2- > V3:
- updated bindings (removed all unnecessary allOf/fixed warnings)
- added dev_err for errors on reading reg and "adi,pll-loop-bandwidth-uhz properties
- replaced hard-coded values with sizeof()

nunojsa commented on drivers/clk/adi/clk-ad9545.c:
```diff
@@ -500,20 +531,45 @@ static int ad9545_parse_dt_plls(struct ad9545_state *st)
+
+			st->pll_clks[addr].profiles[profile_addr].loop_bw_uhz = val;
+
+			ret = fwnode_property_read_u32(profile_node, "adi,profile-priority", &val);
```
Well, my point with `dev_err()` was to use it on all mandatory devicetrees properties... So, for consistency, we should also add it herer and in other aplicable cases...

nunojsa reviewed (commented):
Minor nitpick and I think we are good to go

tachicialex commented:
Changelog V3 -> V4:
- Added errors for all missing properties that are marked as required in bindings/
